Add indeterminate state

boyvanamstel opened the pull request:
Added the indeterminate state from [Carbonize](https://dangercove.com/carbonize) as mentioned in https://github.com/sindresorhus/CircularProgress/issues/9.

~~I need to implement it hiding/showing properly when cancelling the progress, so I marked this as a Draft. (And I wanted to try the new GitHub feature. 🤷‍♂️)~~ All done.

---

Fixes #9

boyvanamstel committed:
Add basic indeterminate state

boyvanamstel commented:
![indeterminate](https://user-images.githubusercontent.com/225410/53012713-18ccb480-3444-11e9-96be-8b21960fd1e1.gif)

boyvanamstel committed:
Properly handle cancelled and finished states

sindresorhus commented:
I don't think the percentage text should show when it's indeterminate.

boyvanamstel commented:
I actually like it more when it's there. That way you know when progress stalled.

boyvanamstel commented:
The example doesn't give the best indication of what that would feel like as it abuses a `Progress` based view and manually activates the indeterminate state.

In a real world scenario it would go up to like 30%, then maybe stall and turn indeterminate. Showing that progress is still at 30%, even though it's currently stalled would be nice to know.

sindresorhus commented:
I think we have different use-cases in mind. I was thinking indeterminate is for when you can't know the progress at all. So if the user just does `.isIndeterminate = true` without setting `.progress`/`progressInstance`, we should not show the progress label. This might be how it currently works. I haven't tested yet.

boyvanamstel commented:
I'll hide the progress label when `.progress == 0` or `.progressInstance == nil`.

boyvanamstel committed:
Hide progress label if progress is 0

boyvanamstel commented:
![indeterminate](https://user-images.githubusercontent.com/225410/53024649-b386bc80-345f-11e9-99df-c117789d73dc.gif)

I've added the animation in `updateLayer()` so it works properly when added and modified through IB. The label is hidden when `progress == nil` or the `progressInstance` hasn't been set and the progress view is set to be indeterminate.

boyvanamstel commented:
Changed my mind about adding the animation in `updateLayer()`. Let's replicate [`startAnimation(_)`](https://developer.apple.com/documentation/appkit/nsprogressindicator/1501167-startanimation) and [`stopAnimation(_)`](https://developer.apple.com/documentation/appkit/nsprogressindicator/1501141-stopanimation) from `NSProgressIndicator`.

I'll update tomorrow.

sindresorhus commented:
> Changed my mind about adding the animation in `updateLayer()`.

I'm curious why you changed your mind?

sindresorhus commented:
This can be done when we've decided on how it should works, but just noting it now while I remember; The readme needs to be updated with the new properties, and also the fact that isIndeterminate keeps the value if triggered when there is progress.

boyvanamstel commented:
> I'm curious why you changed your mind?

I'm not a fan of `if indeterminateCircle.animation(forKey: "rotate") == nil {` being called every time `updateLayer()` is triggered. I tried enabling the animation elsewhere, but everything needs to be added to the view before it'll work. Manually starting and stopping the animation on `NSProgressIndicator` always felt odd to me, but it wouldn't surprise me if this is why.

So I figured why not adopt the same API for `CircularProgress`.

boyvanamstel commented:
> This can be done when we've decided on how it should works, but just noting it now while I remember; The readme needs to be updated with the new properties, and also the fact that isIndeterminate keeps the value if triggered when there is progress.

~~That is on my list, but on halt until the implementation if finalized, like you mentioned.~~
Documentation has been added. 📖

boyvanamstel committed:
Manually start and stop indeterminate animation

boyvanamstel committed:
Update readme

boyvanamstel committed:
Add note about progress and isIndeterminate

sindresorhus commented:
> I'm not a fan of `if indeterminateCircle.animation(forKey: "rotate") == nil {` being called every time `updateLayer()` is triggered.

Why is that a problem? I'd rather have some ugly code internally than having an ugly API externally. So I would prefer not having the `start/stopAnimation` methods.

sindresorhus commented:
> I'll hide the progress label when `.progress == 0` or `.progressInstance == nil`.

You currently hide the progress label when it's higher than 0 too, for example `0.3`.

boyvanamstel commented:
> Why is that a problem? I'd rather have some ugly code internally than having an ugly API externally. So I would prefer not having the `start/stopAnimation` methods.

Maybe that's a personal preference, but I'd rather minimize possible performance problems by exposing them like this. 

I'll revert back to automatically adding the animation though.

boyvanamstel committed:
Revert back to persistently adding animation

sindresorhus commented:
> but I'd rather minimize possible performance problems by exposing them like this.

I added a `print()` to `updateLayer` and let the example app run for a minute and `updateLayer` was called about 30 times. That's less than once a second on average. `.animation(forKey: "rotate")` would have to be extremely slow for that to be an issue.

boyvanamstel committed:
Properly show and hide label when indeterminate

boyvanamstel commented:
<img width="792" alt="screenshot 2019-02-21 at 14 00 49" src="https://user-images.githubusercontent.com/225410/53170764-397e4100-35e1-11e9-9111-5bb25c93c8d4.png">

Added some extra dials and buttons to test this.

sindresorhus commented on Sources/CircularProgress/CircularProgress.swift:
```diff
@@ -74,6 +79,12 @@ public final class CircularProgress: NSView {
+				progressLabel.isHidden = cancelButton.isHidden
+			} else {
+				progressLabel.isHidden = !cancelButton.isHidden
+			}
```
I would prefer a ternary here.

boyvanamstel replied:
I chose against that because it's not the simplest `if`-statement, but if you're ok with it. 👍

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -255,6 +255,16 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
+		self.init()
+		fillColor = nil
+		path = NSBezierPath.circle(radius: radius, center: bounds.center, startAngle: 270.0).cgPath
+		anchorPoint = NSPoint(x: 0.5, y: 0.5)
```
```suggestion
		anchorPoint = CGPoint(x: 0.5, y: 0.5)
```

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -255,6 +255,16 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
 	}
 }
 
+final class IndeterminateShapeLayer: CAShapeLayer {
```
Can you add a short doc comment for this one?

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -361,3 +371,17 @@ extension NSView {
 }
+
+extension CABasicAnimation {
+	static var rotate: CABasicAnimation {
```
Can you add a short doc comment for this one?

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -361,3 +371,17 @@ extension NSView {
+		let animation = CABasicAnimation(keyPath: #keyPath(CAShapeLayer.transform))
+		animation.valueFunction = CAValueFunction(name: .rotateZ)
+		animation.fromValue = 0.0
+		animation.toValue = -(2 * Double.pi)
```
```suggestion
		animation.toValue = -(Double.pi * 2)
```

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -164,13 +164,13 @@ extension NSFont {
 
 extension NSBezierPath {
-	static func circle(radius: Double, center: CGPoint) -> NSBezierPath {
+	static func circle(radius: Double, center: CGPoint, startAngle: CGFloat = 0.0, endAngle: CGFloat = 360.0) -> NSBezierPath {
```
I would prefer to use `Double` here instead of `CGFloat`.

boyvanamstel replied:
I keep forgetting your preference, hah.

sindresorhus commented:
Can you run the GIF through ImageOptim.app?

sindresorhus commented:
This looks great 👌

boyvanamstel committed:
Implement code improvements

boyvanamstel committed:
Run ImageOptim on indeterminate gif

boyvanamstel commented:
> This looks great 👌

Thanks!

boyvanamstel commented:
Rebasing this. Just a sec.

boyvanamstel commented:
@sindresorhus All done. 👍

sindresorhus committed:
Update util.swift

sindresorhus commented:
Looks great!

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -276,6 +276,21 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
+		path = NSBezierPath.circle(radius: radius, center: bounds.center, startAngle: 270).cgPath
+		anchorPoint = CGPoint(x: 0.5, y: 0.5)
+		position = center
+	}
```
@boyvanamstel Would it make sense for this to have a `.isAnimating` property that can start stop the animation?

Then we could move the

```swift
if indeterminateCircle.animation(forKey: "rotate") == nil {
	indeterminateCircle.add(CABasicAnimation.rotate, forKey: "rotate")
}
```

logic here and just do `indeterminateCircle.isAnimating = true` in the `updateColors()` method.

boyvanamstel replied:
It would make it a bit more transparent and easier to handle other animations there later. Might be worth the extra code.

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -276,6 +276,21 @@ final class ProgressCircleShapeLayer: CAShapeLayer {
+
+It draws part of a circle that gets animated into a looping motion around its core.
+*/
+final class IndeterminateShapeLayer: CAShapeLayer {
```
I just realized, shouldn't the name be `IndeterminateShapeLayer` => `IndeterminateCircleShapeLayer`?

boyvanamstel replied:
I thought ‘...CircleShapeLayer’ was the convention because of ‘ProgressCircleShapeLayer’. If it’s not, then ‘IndeterminateShapeLayer’ sounds fine to me. 👍
